pkg/prometheus/promcfg*.go: Validate EnforcedBodySizeLimit

slashpai opened the pull request:
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Added the input validation for EnforcedBodySizeLimit

Related to #4282
## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
Added the input validation for EnforcedBodySizeLimit
```

cc: @simonpasquier

fpetkovski commented on pkg/prometheus/promcfg.go:
```diff
@@ -1804,10 +1804,22 @@ func (l *limitEnforcer) addBodySizeLimitsToYAML(cfg yaml.MapSlice, enforcedLimit
 		return cfg
 	}
 
+	if !validInput("^(0[K,M,G,T,P,E]*B$|[1-9]\\d+[K,M,G,T,P,E]*B$)", enforcedLimit) {
```
Should we use the same library as prometheus [1][2] instead of crafting our own regex? 

[1] https://github.com/prometheus/prometheus/blob/e582e907f8a97ffae27123b9632a254e5018a4a6/config/config.go#L400
[2] https://pkg.go.dev/github.com/alecthomas/units

simonpasquier replied:
Yep using the same library makes sense.

slashpai replied:
I initially created code using units but I dropped that since other fields also needed to be validated even though in this PR only bodysizelimit is validated. So thought to have a general validator method.

slashpai replied:
https://github.com/prometheus-operator/prometheus-operator/compare/master...slashpai:input_validator#diff-453ce9c721c51bf9591c2b212e9ce50157b9c40d1b211177bf502be268099656R1808-R1816

fpetkovski replied:
I think a general validator can be useful as a fallback when more specific ones are not available. Regular expressions can often be hard to verify and troubleshoot, so I would keep things simple if possible.

slashpai replied:
Sure that makes sense, I will change back the validation using units

simonpasquier commented:
I think that the operator should be more explicit and fail the reconciliation if the provided parameter is invalid.

slashpai commented:
> I think that the operator should be more explicit and fail the reconciliation if the provided parameter is invalid.

I propagated the error from validation that should fail sync

slashpai commented:
@fpetkovski @simonpasquier Can you review again?

fpetkovski commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -5856,9 +5882,6 @@ alerting:
 				nil,
 				nil,
 			)
-			if err != nil {
```
Maybe we want to refine this assertion like 
```
expectErr := expected == ""
if expectErr && err == nil &&  {
  t.Fatal("prometheus config generation should return an error")
}

if !expectErr && err != nil {
  t.Fatal("prometheus config generation should not return an error")
}
```
Alternatively, we could also add a `expectErr` field to the test cases and use it instead of relying on the value of `expected`

slashpai replied:
@fpetkovski Thank you :) . I updated the test

fpetkovski commented:
Nice work, this LGMT!

simonpasquier reviewed (commented):
IMHO the check should happen near the top of `*ConfigGenerator.GenerateConfig()` since it's a PrometheusSpec field?

https://github.com/prometheus-operator/prometheus-operator/blob/c6b5cc2e178002683ab3afe8d023a8bb2508ba19/pkg/prometheus/promcfg.go#L219

slashpai commented:
> IMHO the check should happen near the top of `*ConfigGenerator.GenerateConfig()` since it's a PrometheusSpec field?
> 
> https://github.com/prometheus-operator/prometheus-operator/blob/c6b5cc2e178002683ab3afe8d023a8bb2508ba19/pkg/prometheus/promcfg.go#L219

Good point. I added a func for validating the fields at prometheus crd level. Added the func call in operator.go before calling GenerateConfig

@simonpasquier Can you take a look?

simonpasquier commented on pkg/prometheus/operator.go:
```diff
@@ -1592,6 +1592,13 @@ func (c *Operator) createOrUpdateConfigurationSecret(ctx context.Context, p *mon
 	}
 
+	// Validate Prometheus Config Inputs at Prometheus CRD level
+	err = c.configGenerator.ValidateConfigInputs(p)
```
I'm not sure we need to create another receiver method for `*ConfigGenerator`. IMHO the validation should be done by `GenerateConfig()` directly so the caller doesn't have to worry about validating th configuration before generating.

slashpai replied:
Did you mean to call the ValidateInput() func from inside GenerateConfig() method?

simonpasquier replied:
yes and we don't need to make `ValidateConfigInputs()` public (e.g. `validateConfigInputs()` instead).

slashpai replied:
thank you made it public since I was trying to call from operator.go. I am updating with changes

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -215,6 +216,31 @@ func buildExternalLabels(p *v1.Prometheus) yaml.MapSlice {
 
+// ValidateConfigInputs validates if inputs passed are valid
+func (cg *ConfigGenerator) ValidateConfigInputs(p *v1.Prometheus) error {
+	logger := log.With(cg.logger, "Validating config inputs", "EnforcedBodySizeLimit")
```
Since the logger isn't used directly in this func, no need to create a new contextual logger. And the "key" parameter should be snake case instead of a full sentence.

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -215,6 +216,31 @@ func buildExternalLabels(p *v1.Prometheus) yaml.MapSlice {
+	_, err := units.ParseBase2Bytes(enforcedLimit)
+
+	if err != nil {
+		level.Warn(logger).Log("msg", "body_size_limit provided is invalid it should be a format of size",
+			"enforcedBodySizeLimit", enforcedLimit)
+		return errors.Wrap(err, "Invalid body_size_limit value specified")
```
In general you should either return an error or log a message but not doing both.
See "Only handle errors once" in https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully. And the rest of the post is only worth reading ;-)

```suggestion
		return errors.Wrap(err, "invalid enforcedBodySizeLimit value specified")
```

simonpasquier reviewed (commented):
Can you base your change against release-0.51 so it can be part of v0.51.2

slashpai commented:
@simonpasquier I updated with changed and changed the base for PR

slashpai commented:
> @simonpasquier I updated with changed and changed the base for PR
just realized I didn't update error string, updating soon

slashpai commented:
Done

slashpai commented:
@simonpasquier Take a look when you get chance

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -227,6 +250,13 @@ func (cg *ConfigGenerator) GenerateConfig(
 	ruleConfigMapNames []string,
 ) ([]byte, error) {
+	// Validate Prometheus Config Inputs at Prometheus CRD level
+	err := validateConfigInputs(p)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid config value passed")
+	}
```
I think we can avoid wrapping the error, the returned error should already be clear enough.

```suggestion
	if err := validateConfigInputs(p); err != nil {
		return nil, err
	}
```

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -215,6 +216,28 @@ func buildExternalLabels(p *v1.Prometheus) yaml.MapSlice {
+	return nil
+}
+
+// Validate EnforcedBodySizeLimit
```
```suggestion
```

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -215,6 +216,28 @@ func buildExternalLabels(p *v1.Prometheus) yaml.MapSlice {
 	return stringMapToMapSlice(m)
 }
 
+// ValidateConfigInputs validates if inputs passed are valid
```
```suggestion
// validateConfigInputs runs extra validation on the Prometheus fields which can't be done at the CRD schema validation level.
```

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -215,6 +216,28 @@ func buildExternalLabels(p *v1.Prometheus) yaml.MapSlice {
+// ValidateConfigInputs validates if inputs passed are valid
+func validateConfigInputs(p *v1.Prometheus) error {
+	if p.Spec.EnforcedBodySizeLimit != "" {
+		err := validateBodySizeLimit(p.Spec.EnforcedBodySizeLimit)
+		if err != nil {
+			return err
+		}
```
```suggestion
		if err := validateBodySizeLimit(p.Spec.EnforcedBodySizeLimit); err != nil {
			return err
		}
```

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -215,6 +216,28 @@ func buildExternalLabels(p *v1.Prometheus) yaml.MapSlice {
+// Validate EnforcedBodySizeLimit
+func validateBodySizeLimit(enforcedLimit string) error {
+	// To validate if given value is parsable for the acceptable body_size_limit values
+	_, err := units.ParseBase2Bytes(enforcedLimit)
+
+	if err != nil {
+		return errors.Wrap(err, "invalid enforcedBodySizeLimit value specified")
+	}
```
```suggestion
	if _, err := units.ParseBase2Bytes(enforcedLimit); err != nil {
		return errors.Wrap(err, "invalid enforcedBodySizeLimit value specified")
	}
```

slashpai commented:
@simonpasquier Thank you for the review. I updated the PR. Hope its good now :)

simonpasquier reviewed (approved):
lgtm

slashpai commented:
@simonpasquier Curious to know would this change be synced with master only before release of 0.51.2?

simonpasquier commented:
@slashpai yes this is the process.
